include/operations/SymbolicOperation.hpp

```diff
@@ -0,0 +1,129 @@
+#pragma once
+
+#include "Definitions.hpp"
+#include "Expression.hpp"
+#include "StandardOperation.hpp"
+
+#include <algorithm>
+#include <cstddef>
+#include <optional>
+#include <unordered_set>
+#include <variant>
+
+namespace qc {
+    // Overload pattern for std::visit
+    template<typename... Ts>
+    struct Overload: Ts... {
+        using Ts::operator()...;
+    };
+    template<class... Ts>
+    Overload(Ts...) -> Overload<Ts...>;
+
+    class SymbolicOperation final: public StandardOperation {
+    protected:
+        std::array<std::optional<Symbolic>, MAX_PARAMETERS> symbolicParameter{{std::nullopt, std::nullopt, std::nullopt}};
+
+        static OpType parseU3(const Symbolic& lambda, dd::fp& phi, dd::fp& theta);
+        static OpType parseU3(dd::fp& lambda, const Symbolic& phi, dd::fp& theta);
+        static OpType parseU3(dd::fp& lambda, dd::fp& phi, const Symbolic& theta);
+        static OpType parseU3(const Symbolic& lambda, const Symbolic& phi, dd::fp& theta);
+        static OpType parseU3(const Symbolic& lambda, dd::fp& phi, const Symbolic& theta);
+        static OpType parseU3(dd::fp& lambda, const Symbolic& phi, const Symbolic& theta);
+
+        static OpType parseU2(const Symbolic& lambda, const Symbolic& phi);
+        static OpType parseU2(const Symbolic& lambda, dd::fp& phi);
+        static OpType parseU2(dd::fp& lambda, const Symbolic& phi);
+
+        static OpType parseU1(const Symbolic& lambda);
+
+        void checkUgate();
+
+        void storeSymbolOrNumber(const SymbolOrNumber& param, std::size_t i);
+
+        [[nodiscard]] bool isSymbolicParameter(std::size_t i) const {
+            return symbolicParameter[i].has_value();
+        }
+
+        static bool isSymbol(const SymbolOrNumber& param) {
+            return std::holds_alternative<Symbolic>(param);
+        }
+
+        static Symbolic& getSymbol(SymbolOrNumber& param) {
+            return std::get<Symbolic>(param);
+        }
+
+        static dd::fp& getNumber(SymbolOrNumber& param) {
+            return std::get<dd::fp>(param);
+        }
+
+        void setup(dd::QubitCount nq, const SymbolOrNumber& par0, const SymbolOrNumber& par1, const SymbolOrNumber& par2, dd::Qubit startingQubit = 0);
+
+        [[nodiscard]] static dd::fp getInstantiation(const SymbolOrNumber& symOrNum, const VariableAssignment& assignment);
+
+    public:
+        SymbolicOperation() = default;
+
+        // [[nodiscard]] const std::array<Symbolic, MAX_PARAMETERS>& getSymbolicParameter() const {
+        //     return symbolicParameter;
+        // }
+
+        // std::array<Symbolic, MAX_PARAMETERS>& getSymbolicParameter() {
+        //     return symbolicParameter;
+        // }
+
+        [[nodiscard]] SymbolOrNumber getParameter(std::size_t i) const {
+            if (symbolicParameter[i].has_value()) return symbolicParameter[i].value();
+            return parameter[i];
+        }
+
+        void setSymbolicParameter(const Symbolic& par, std::size_t i) {
```

Comment:
```suggestion
        void setSymbolicParameter(const Symbolic& par, const std::size_t i) {
```

Reply:
I'm thinking about making an emacs shortcout for just typing const.

Reply:
Just do it ✅